Mathlib/Tactic/Linter/Header.lean

```diff
@@ -0,0 +1,296 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Rothgang, Damiano Testa
+-/
+import Lean.Elab.Command
+
+/-!
+#  The "header" linter
+
+The "header" style linter checks that a file starts with
+```
+/-
+Copyright ...
+Apache ...
+Authors ...
+-/
+
+import statements*
+module doc-string*
+remaining file
+```
+It emits a warning if
+* the copyright statement is malformed;
+* `Mathlib.Tactic` is imported;
+* any import in `Lake` is present;
+* the first non-`import` command is not a module doc-string.
+
+The linter allows `import`-only files and does not require a copyright statement in `Mathlib.Init`.
+
+## Implementation
+The strategy used by the linter is as follows.
+The linter computes the end position of the first module doc-string of the file,
+resorting to the end of the file, if there is no module doc-string.
+Next, the linter tries to parse the file up to the position determined above.
+
+If the parsing is successful, the linter checks the resulting `Syntax` and behaves accordingly.
+
+If the parsing is not successful, this already means there is some "problematic" command
+after the imports. In particular, there is a command that is not a module doc-string
+immediately following the last import: the file should be flagged by the linter.
+Hence, the linter then falls back to parsing the header of the file, adding a spurious `section`
+after it.
+This makes it possible for the linter to check the entire header of the file, emit warnings that
+could arise from this part and also flag that the file should contain a module doc-string after
+the `import` statements.
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter
+
+/--
+`firstNonImport? stx` assumes that the input `Syntax` is of kind `Lean.Parser.Module.module`.
+It returns
+* `none`, if `stx` consists only of `import` statements,
+* the first non-`import` command in `stx`, otherwise.
+
+The intended use-case is to use the output of `testParseModule` as the input of
+`firstNonImport?`.
+-/
+def firstNonImport? : Syntax → Option Syntax
+  | .node _ ``Lean.Parser.Module.module #[_header, .node _ `null args] => args[0]?
+  | _=> some .missing  -- this is unreachable, if the input comes from `testParseModule`
+
+/-- `getImportIds s` takes as input `s : Syntax`.
+It returns the array of all `import` identifiers in `s`. -/
+partial
+def getImportIds (s : Syntax) : Array Syntax :=
+  let rest : Array Syntax := (s.getArgs.map getImportIds).flatten
+  if s.isOfKind ``Lean.Parser.Module.import then
+    rest.push (s.getArgs.getD 2 default)
+  else
+    rest
+
+/--
+`parseUpToHere pos post` takes as input `pos : String.Pos` and the optional `post : String`.
+It parses the current file from the beginning until `pos`, appending `post` at the end.
+
```

Comment:
The following fact is used to make `firstNonImport?` work later: how about mentioning it?
```suggestion
It returns a syntax node of kind `Lean.Parser.Module.module`.